src/asl/synth.ts

```diff
@@ -2265,7 +2269,11 @@ export class ASL {
             prop: PropAssignExpr | SpreadAssignExpr
           ): ASLGraph.LiteralValue | ASLGraph.JsonPath {
             const output = evalExprToJsonPathOrLiteral(prop.expr);
-            return ASLGraph.isJsonPath(output) ? assignValue(output) : output;
+            return ASLGraph.isJsonPath(output) &&
+              // paths at $$ are immutable
+              !output.jsonPath.startsWith("$$.")
+              ? assignValue(output)
+              : output;
```

Comment:
I don't understand the significance of this branch

Reply:
`evalSafeOutput` references jsonPath values to ensure they do not change between the time when they are supposed to be set in javascript and when they are actually set in stepfunctions.

```ts
let x = 0;
{
    a: x,
    b: (x = 1)
}
``` 

If we don't reference x, it may be 1 at `a` and not 0.

So we do...

```ts
let x = 0;
let v1 = x;
{
    a: v1,
    b: (x = 1)
}
``` 

However, `$$` json paths are the step functions context object which cannot be assigned by the user (and the values we expose do not change between states).

It is safe to use them directly.

```ts
let x = 0;
{
    z: context.Execution.Name,
    a: x,
    b: (x = 1)
}
``` 

becomes  (we don't change the AST, using it for illustration)

```ts
let x = 0;
let v1 = x;
{
    z: context.Execution.Name,
    a: v1,
    b: (x = 1)
}
```